src/objective/regression_obj.cc

```diff
@@ -38,27 +38,29 @@ class RegLossObj : public ObjFunction {
       const std::vector<std::pair<std::string, std::string> > &args) override {
     param_.InitAllowUnknown(args);
   }
-  void GetGradient(const std::vector<bst_float> &preds, const MetaInfo &info,
-                   int iter, std::vector<bst_gpair> *out_gpair) override {
+  void GetGradient(HostDeviceVector<bst_float> *preds, const MetaInfo &info,
+                   int iter, HostDeviceVector<bst_gpair> *out_gpair) override {
     CHECK_NE(info.labels.size(), 0U) << "label set cannot be empty";
-    CHECK_EQ(preds.size(), info.labels.size())
+    CHECK_EQ(preds->size(), info.labels.size())
         << "labels are not correctly provided"
-        << "preds.size=" << preds.size()
+        << "preds.size=" << preds->size()
         << ", label.size=" << info.labels.size();
+    auto& preds_h = preds->data_h();
 
     this->LazyCheckLabels(info.labels);
-    out_gpair->resize(preds.size());
-    const omp_ulong n = static_cast<omp_ulong>(preds.size());
-    auto gpair_ptr = out_gpair->data();
+    out_gpair->resize(preds_h.size(), -1);
+    auto& gpair = out_gpair->data_h();
+    const omp_ulong n = static_cast<omp_ulong>(preds_h.size());
+    auto gpair_ptr = out_gpair->ptr_h();
     avx::Float8 scale(param_.scale_pos_weight);
 
     const omp_ulong remainder = n % 8;
     int nthread = omp_get_max_threads();
     // Use a maximum of 8 threads
 #pragma omp parallel for schedule(static) num_threads(std::min(8, nthread))
-    for (omp_ulong i = 0; i < n - remainder; i += 8) {
+    for (int i = 0; i < n - remainder; i += 8) {
```

Comment:
Can we keep the loop variable to `omp_ulong`? Any reason to change?

Reply:
Done.